### PR TITLE
Fix return type of linalg.norm for complex input

### LINKED ISSUE #780: cupy.linalg.norm returns complex scalar for complex input
```python
>>> a = cupy.array([1j, 2, 3])
>>> a.dtype
dtype('complex128')
>>> cupy.linalg.norm(a).dtype
dtype('complex128')
```
It should be `float64` in this case.

### PR DESCRIPTION
**Pull request by kohr-h**
Closes #780

### CODE REVIEW

File: cupy/linalg/norms.py [hunk 1]
```diff
@@ -70,14 +75,28 @@ def norm(x, ord=None, axis=None, keepdims=False):
...
         else:
             try:
                 float(ord)
             except TypeError:
                 raise ValueError("Invalid norm order for vectors.")
-            absx = abs(x).astype('d')
+
+            # Mirror Numpy behavior of casting to double for non-complex
+            # dtypes or reduction over all axes
+            reduce_all_axes = (tuple(sorted(axis)) == tuple(range(nd)) and
```
Comment (okuta) [inline cupy/linalg/norms.py]: How about moving after L94?

Comment (kohr-h) [reply cupy/linalg/norms.py]: Sorry, I don't quite understand what you mean. Do you want me to inline the definition in line 95?

Comment (okuta) [reply cupy/linalg/norms.py]: Sorry about less information.
```python
cast_dtype = 'd'
if issubclass(x.dtype.type, numpy.complexfloating):
    # Mirror Numpy behavior of casting to double for non-complex
    # dtypes or reduction over all axes
    if keepdims or tuple(sorted(axis)) != tuple(range(nd):
       cast_dtype = x.dtype.char.lower()  # 'D'->'d' and 'F'->'f'
```

Comment (kohr-h) [reply cupy/linalg/norms.py]: Thanks for the clarification, I implemented the change.

**Comment by okuta** [comment]
Thanks.
jenkins, test this please

**Comment by kohr-h** [comment]
Ready to merge?

**Comment by okuta** [comment]
LGTM!

**Pull request merged**